router/src/completion.rs

```diff
@@ -0,0 +1,581 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+use crate::{default_max_new_tokens, GenerateRequest, GenerateParameters, FinishReason, GenerateResponse, Info, StreamType, StreamResponse};
+use axum::Json;
+use axum::extract::Extension;
+use erased_serde;
+
+#[derive(Clone, Debug, Deserialize, ToSchema)]
+pub(crate) struct CompatCompletionRequest {
+    #[schema(example = "My name is Michael and I")]
+    pub prompt: String,
+    #[serde(default)]
+    #[schema(exclusive_minimum = 0, nullable = true, default = "null", example = 1)]
+    pub best_of: Option<usize>,
+    #[serde(default)]
+    #[schema(
+        exclusive_minimum = 0.0,
+        nullable = true,
+        default = "null",
+        example = 0.5
+    )]
+    pub temperature: Option<f32>,
+    #[serde(default)]
+    #[schema(
+        exclusive_minimum = -2.0,
+        nullable = true,
+        default = "null",
+        example = 0.0
+    )]
+    pub presence_penalty: Option<f32>,
+    // #[serde(default)]
+    // #[schema(exclusive_minimum = 0, nullable = true, default = 1, example = 1)]
+    // pub n: Option<i32>,
+    #[serde(default)]
+    #[schema(exclusive_minimum = 0, nullable = true, default = "null", example = 10)]
+    pub top_k: Option<i32>,
+    #[serde(default)]
+    #[schema(
+        exclusive_minimum = 0.0,
+        maximum = 1.0,
+        nullable = true,
+        default = "null",
+        example = 0.95
+    )]
+    pub top_p: Option<f32>,
+    #[serde(default)]
+    #[schema(
+        exclusive_minimum = 0.0,
+        maximum = 1.0,
+        nullable = true,
+        default = "null",
+        example = 0.95
+    )]
+    pub typical_p: Option<f32>,
+    #[serde(default)]
+    #[schema(default = "false", example = true)]
+    pub do_sample: bool,
+    #[serde(default = "default_max_new_tokens")]
+    #[schema(exclusive_minimum = 0, exclusive_maximum = 512, default = "20")]
+    pub max_tokens: u32,
+    #[serde(default)]
+    #[schema(nullable = true, default = "null", example = false)]
+    pub echo: Option<bool>,
+    #[serde(default)]
+    #[schema(inline, max_items = 4, example = json ! (["photographer"]))]
+    pub stop: Vec<String>,
+    #[serde(default)]
+    #[schema(nullable = true, default = "null", example = "null")]
+    pub truncate: Option<usize>,
+    #[serde(default)]
+    #[schema(default = "false", example = true)]
+    pub watermark: bool,
+    #[serde(default)]
+    #[schema(default = "false")]
+    pub decoder_input_details: bool,
+    #[serde(default)]
+    #[schema(
+        exclusive_minimum = 0,
+        nullable = true,
+        default = "null",
+        example = "null"
+    )]
+    pub seed: Option<u64>,
+    #[serde(default)]
+    #[schema(default = "false")]
+    pub stream: bool,
+}
+
+
+impl From<CompatCompletionRequest> for GenerateRequest {
```

Comment:
Why do we use `From` instead of an explicitly called conversion function? Why do we ever need a converter here?

Reply:
Same architecture choice as here: https://github.com/Preemo-Inc/text-generation-inference/blob/d2ae3581bf17749c5fbaf1ae6812aa158ffeccca/router/src/lib.rs#L184